src/clangd-context.ts

```diff
@@ -66,6 +66,10 @@ export class ClangdContext implements vscode.Disposable {
     if (!clangdPath)
       return;
 
+    if (!config.get<boolean>('enable')) {
```

Comment:
Any reason not to move the check even earlier (in particular before the `install.activate` call, but maybe even at the top of the `activate` function in `extension.ts`, thereby avoiding the need for the added check there)?

Reply:
No reason except my lack of familiarity with the code. Thanks for the feedback! I'll give this a go.

Reply:
I tried this out and it's a bit more complicated than I expected. In `extension.ts` I can't return early because `activate` needs to return a `ClangdExtension` so at least some of that code needs to run to create and configure the context used to create the `ClangdExtensionImpl`. I think the safest simplification I can make that doesn't result in a half-initialized context is to do this in `extension.ts`:

```javascript
  let shouldCheck = false;

  if (!vscode.workspace.getConfiguration('clangd').get<boolean>('enable')) {
    await clangdContext.activate(context.globalStoragePath, outputChannel);

    shouldCheck = vscode.workspace.getConfiguration('clangd').get<boolean>(
      'detectExtensionConflicts') ?? false;
  }

  if (shouldCheck) {
    const interval = setInterval(function() {
      ...
```

Would you like me to make this change?

Reply:
This actually raises an interesting issue. The reason `activate()` returns anything is that we expose an [API](https://github.com/clangd/vscode-clangd/blob/master/api/README.md) for use by other extensions. So the question is: how should the API behave if `clangd.enable` is set to `false`.

With the current patch, it looks like we pass a `null` client to `ClangdExtension`. This is technically not permitted by the API, since the client object [it exposes](https://github.com/clangd/vscode-clangd/blob/4f64ca7d426e36620164a635f679659d471f21a4/api/vscode-clangd.d.ts#L11) has type `BaseLanguageClient`, not `BaseLanguageClient | null`. However, it type-checks because we [fudge the check](https://github.com/clangd/vscode-clangd/blob/4f64ca7d426e36620164a635f679659d471f21a4/src/clangd-context.ts#L61) on the nullness of `ClangdContext.client` itself.

I guess we [already have a codepath](https://github.com/clangd/vscode-clangd/blob/4f64ca7d426e36620164a635f679659d471f21a4/src/clangd-context.ts#L66-L67) that can leave `ClangdContext.client` as null, so adding another one is fine for now. But we should revisit this in a follow-up issue and consider changing the API to explicitly make the exposed client object nullable (or maybe allow a null `ClangdExtension` return value for `activate` itself).

TL;DR: Yes, please go ahead and make the described simplification.

Reply:
> we should revisit this in a follow-up issue and consider changing the API to explicitly make the exposed client object nullable (or maybe allow a null `ClangdExtension` return value for `activate` itself).

I filed https://github.com/clangd/vscode-clangd/issues/721 about this.

Reply:
Thanks again for your help. I rebased, applied the simplified change, and re-tested this.